python/pyarrow/src/arrow/python/numpy_interop.h

```diff
@@ -67,6 +67,13 @@
 #define NPY_INT32_IS_INT 0
 #endif
 
+// Backported NumPy 2 API (can be removed if numpy 2 is required)
+#if NPY_ABI_VERSION < 0x02000000
+#define PyDataType_ELSIZE(descr) ((descr)->elsize)
+#define PyDataType_C_METADATA(descr) ((descr)->c_metadata)
+#define PyDataType_FIELDS(descr) ((descr)->fields)
+#endif
```

Comment:
Can we use the following until NumPy's Windows nightly wheel build is fixed?

```suggestion
#ifndef PyDataType_ELSIZE
#define PyDataType_ELSIZE(descr) ((descr)->elsize)
#endif
#ifndef PyDataType_C_METADATA
#define PyDataType_C_METADATA(descr) ((descr)->c_metadata)
#endif
#ifndef PyDataType_FIELDS
#define PyDataType_FIELDS(descr) ((descr)->fields)
#endif
```

Reply:
Hmmm, I guess it bites us that we don't have more fine grained patch versions :(.  Those are defined as static inline functions unfortunately.  You could put `#ifndef _PyArray_DescrNumPy2`, but it would be a temporary work-around :(.

Reply:
Sorry, too late, that doesn't make sense either... It is a struct after all.  I don't think there is a way to distinguish an older nightly :(.

Reply:
Ah, sorry. I should have checked how to implement these APIs.